File: src/main/java/nl/chimpgamer/networkmanager/extensions/discordbot/tasks/VerifyUserTask.java
package nl.chimpgamer.networkmanager.extensions.discordbot.tasks;

import com.google.common.base.Preconditions;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.Role;
import nl.chimpgamer.networkmanager.api.models.player.Player;
import nl.chimpgamer.networkmanager.common.utils.Methods;
import nl.chimpgamer.networkmanager.extensions.discordbot.DiscordBot;
import nl.chimpgamer.networkmanager.extensions.discordbot.api.events.PlayerVerifyEvent;
import nl.chimpgamer.networkmanager.extensions.discordbot.api.models.Token;
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.Setting;
import nl.chimpgamer.networkmanager.extensions.discordbot.manager.DiscordUserManager;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.DCMessage;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.JsonEmbedBuilder;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.MCMessage;
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.Utils;

import java.sql.SQLException;

@Getter
@RequiredArgsConstructor
public class VerifyUserTask implements Runnable {
    private final DiscordBot discordBot;
    private final Player player;
    private final Token token;

    @Override
    public void run() {
        if (this.getToken() == null) {
            this.getPlayer().sendMessage(MCMessage.VERIFY_INVALID_TOKEN.getMessage()
                    .replace("%playername%", this.getPlayer().getName()));
            return;
        }

        if (this.getToken().getCreated() + 300000 < System.currentTimeMillis()) {
            // Token Expired
            this.getPlayer().sendMessage(MCMessage.VERIFY_TOKEN_EXPIRED.getMessage()
                    .replace("%playername%", this.getPlayer().getName()));
        } else {
            try {
                Preconditions.checkNotNull(getDiscordBot().getGuild(), "The discord bot has not been connected to a discord server. Connect it to a discord server.");

                DiscordUserManager discordUserManager = this.getDiscordBot().getDiscordUserManager();

                Member member = getDiscordBot().getGuild().getMemberById(this.getToken().getDiscordID());
                if (member == null) {
                    this.getPlayer().sendMessage(MCMessage.VERIFY_HELP.getMessage());
                    return;
                }
                if (discordUserManager.checkUser(this.getPlayer().getUuid().toString())) {
                    // User is already registered...
                    this.getPlayer().sendMessage(MCMessage.VERIFY_ACCOUNT_ALREADY_LINKED.getMessage()
                            .replace("%playername%", this.getPlayer().getName()));
                } else {
                    // User is not registered yet...
                    discordUserManager.getTokens().remove(this.getToken()); // Remove token
                    discordUserManager.insertUser(this.getPlayer().getUuid(), this.getToken().getDiscordID());

                    String registrationCompleted = DCMessage.REGISTRATION_COMPLETED.getMessage();
                    if (Methods.isJsonValid(registrationCompleted)) {
                        JsonEmbedBuilder jsonEmbedBuilder = JsonEmbedBuilder.fromJson(registrationCompleted);
                        Utils.editMessage(this.getToken().getMessage(), jsonEmbedBuilder.build());
                    } else {
                        Utils.editMessage(this.getToken().getMessage(), registrationCompleted);
                    }

                    this.getPlayer().sendMessage(MCMessage.VERIFY_COMPLETED.getMessage()
                            .replace("%playername%", this.getPlayer().getName()));
                    this.getDiscordBot().getNetworkManager().getEventHandler().callEvent(new PlayerVerifyEvent(this.getPlayer(), member));
                    if (this.getDiscordBot().getNetworkManager().isRedisBungee()) {
                        this.getDiscordBot().sendRedisBungee("load " + this.getPlayer().getUuid());
                    }

                    if (Setting.DISCORD_VERIFY_ADD_ROLE_ENABLED.getAsBoolean()) {
                        Role verifiedRole = this.getDiscordBot().getDiscordManager().getVerifiedRole();
                        if (verifiedRole != null) {
                            this.getDiscordBot().getLogger().info("Assigning the " + verifiedRole.getName() + " role to " + member.getEffectiveName());
                            Utils.addRoleToMember(member, verifiedRole);
                        }
                    }

                    if (Setting.DISCORD_SYNC_USERNAME.getAsBoolean()) {
                        Utils.setNickName(member, this.getPlayer().getName());
                    }

                    if (Setting.DISCORD_SYNC_RANKS_ENABLED.getAsBoolean()) {
                        Utils.syncRanks(this.getPlayer());
                    }
                }
            } catch (SQLException ex) {
                this.getPlayer().sendMessage(MCMessage.VERIFY_ERROR.getMessage()
                        .replace("%playername%", this.getPlayer().getName()));
                ex.printStackTrace();
            }
        }
    }
}

File: src/main/java/nl/chimpgamer/networkmanager/extensions/discordbot/tasks/VerifyUserTask.kt
package nl.chimpgamer.networkmanager.extensions.discordbot.tasks

import com.google.common.base.Preconditions
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Role
import nl.chimpgamer.networkmanager.api.models.player.Player
import nl.chimpgamer.networkmanager.common.utils.Methods
import nl.chimpgamer.networkmanager.extensions.discordbot.DiscordBot
import nl.chimpgamer.networkmanager.extensions.discordbot.api.events.PlayerRegisteredEvent
import nl.chimpgamer.networkmanager.extensions.discordbot.api.models.Token
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.DCMessage
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.MCMessage
import nl.chimpgamer.networkmanager.extensions.discordbot.configurations.Setting
import nl.chimpgamer.networkmanager.extensions.discordbot.manager.DiscordUserManager
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.JsonEmbedBuilder
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.Utils
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.Utils.addRoleToMember
import nl.chimpgamer.networkmanager.extensions.discordbot.utils.Utils.setNickName
import java.sql.SQLException

class VerifyUserTask(private val discordBot: DiscordBot, private val player: Player, private val token: Token) : Runnable {

    override fun run() {
        if (this.token.created + 300000 < System.currentTimeMillis()) { // Token Expired
            this.player.sendMessage(MCMessage.REGISTER_TOKEN_EXPIRED.message
                    .replace("%playername%", this.player.name))
        } else {
            try {
                Preconditions.checkNotNull(this.discordBot.guild, "The discord bot has not been connected to a discord server. Connect it to a discord server.")
                val discordUserManager: DiscordUserManager = this.discordBot.discordUserManager
                val member: Member? = this.discordBot.guild.getMemberById(this.token.discordID)
                if (member == null) {
                    this.player.sendMessage(MCMessage.REGISTER_NOT_IN_SERVER.message)
                    return
                }
                if (discordUserManager.existsInDatabase(this.player.uuid.toString())) { // User is already registered...
                    this.player.sendMessage(MCMessage.REGISTER_ACCOUNT_ALREADY_LINKED.message
                            .replace("%playername%", this.player.name))
                } else { // User is not registered yet...
                    discordUserManager.tokens.remove(this.token) // Remove token
                    discordUserManager.insertUser(this.player.uuid, this.token.discordID)
                    val registrationCompleted = DCMessage.REGISTRATION_COMPLETED.message
                    if (Methods.isJsonValid(registrationCompleted)) {
                        val jsonEmbedBuilder = JsonEmbedBuilder.fromJson(registrationCompleted)
                        Utils.editMessage(this.token.message, jsonEmbedBuilder.build())
                    } else {
                        Utils.editMessage(this.token.message, registrationCompleted)
                    }
                    this.player.sendMessage(MCMessage.REGISTER_COMPLETED.message
                            .replace("%playername%", this.player.name))
                    this.discordBot.networkManager.eventHandler.callEvent(PlayerRegisteredEvent(this.player, member))
                    if (this.discordBot.networkManager.isRedisBungee) {
                        this.discordBot.sendRedisBungee("load " + this.player.uuid)
                    }
                    if (Setting.DISCORD_REGISTER_ADD_ROLE_ENABLED.asBoolean) {
                        val verifiedRole: Role? = this.discordBot.discordManager.verifiedRole
                        if (verifiedRole != null) {
                            this.discordBot.logger.info("Assigning the " + verifiedRole.name + " role to " + member.effectiveName)
                            addRoleToMember(member, verifiedRole)
                        }
                    }
                    if (Setting.DISCORD_SYNC_USERNAME.asBoolean) {
                        setNickName(member, this.player.name)
                    }
                    if (Setting.DISCORD_SYNC_RANKS_ENABLED.asBoolean) {
                        discordBot.scheduler.runDelayed(SyncRanksTask(discordBot, player), 1)
                    }
                }
            } catch (ex: SQLException) {
                this.player.sendMessage(MCMessage.REGISTER_ERROR.message
                        .replace("%playername%", this.player.name))
                ex.printStackTrace()
            }
        }
    }
}